### PR TITLE
Allow config variables to provide conversion functions

### PR DESCRIPTION
**Pull request by jamesls**
This is primarily to address #395, where the `metadata_*` variables need to be converted to integers.

To fix this, the session var mappings now have an additional field, which is a conversion function, which if provided, will be called before returning the value in `get_config_variable`.

The main change is provided in https://github.com/boto/botocore/commit/b0ea7bc62550665488a69ca0beeb9a240f249c50.  I've also included a number of commits for code cleanup to the session module.

Additionally, I've made one change to the signature of `get_config_variable` which is to remove the `default` argument.  This simplifies the code logic, and nothing in botocore, the cli, nor boto3 used this functionality.  I'd prefer to remove this until there's a strong case for adding this back.  It also simplifies the code logic.

This will also require a corresponding change to the CLI.

cc @kyleknap @mtdowling

### CODE REVIEW

File: botocore/session.py [hunk 1]
```diff
@@ -53,7 +49,9 @@ class Session(object):
     :ivar profile: The current profile.
     """
 
-    AllEvents = {
+    #: A dictionary where each key is an event name and the value
+    #: is the formatting string used to construct a new event.
+    ALL_EVENTS = {
```
Comment (kyleknap) [inline botocore/session.py]: The event names look out of date as well.

File: docs/source/index.rst [hunk 2]
```diff
@@ -35,6 +35,14 @@ Contents:
 Upgrade Notes
 =============
 
+Upgrading to 1.0.0rc1
```
Comment (kyleknap) [inline docs/source/index.rst]: I forgot that we started this. There may be some things that I forgot to add like the update to the pagination interface, but I am not entirely sure when it was added.

**Comment by kyleknap** [comment]
LGTM. Had a couple comments that were sort of unrelated to the PR, but :ship: otherwise.

File: docs/source/index.rst (hunk 2 above)
Comment (kyleknap) [inline docs/source/index.rst]: Should we backport them in another PR?

Comment (jamesls) [inline docs/source/index.rst]: Yeah, it might be worth backporting these in another PR.

File: botocore/session.py (hunk 1 above)
Comment (jamesls) [inline botocore/session.py]: Yeah, this is interesting because as we started to refactor code to be less dependent on a session, and instead just take event_emitters where necessary, they now no longer have this list of available events accessible.

The original idea behind this was to a) see what events were available and b) avoid typos in `emit()/register()` calls when referencing event names.

What do you think we should do with these now?

Comment (kyleknap) [inline botocore/session.py]: I do not even sure what they are used for as they are inaccurate. If the functionality is not really used anymore, I would be in favor of removing the global dictionary and any methods that use the dictionary.

Comment (jamesls) [inline botocore/session.py]: Ok, I'll remove them.

**Comment by jamesls** [comment]
@kyleknap Feedback incorporated, mind taking another look?

**Comment by kyleknap** [comment]
@jamesls 
Looks like you need to rebase. You removed an import that is needed for another test.

**Comment by kyleknap** [comment]
Otherwise, once the test pass. You are good to merge. :ship:
